scripts/tools/not_known_to_gn.py

```diff
@@ -0,0 +1,179 @@
+        #     upper directory (although upper directory is not ideal)
+        for p in file.parents:
+            data = self.gn_data.get(str(p), None)
+            if not data:
+                continue
+
+            if file.name in data:
+                logging.debug("%s found in BUILD.gn for %s", file, p)
+                return
+
+        path = str(file.relative_to(top_dir))
+        if not self._IsKnownFailure(path):
+            logging.error("UNKNOWN to gn: %s", path)
+            self.fatal_failures += 1
+        else:
+            logging.warning("UNKNOWN to gn: %s (known error)", path)
+
+        self.failures += 1
+
+
+@click.command()
+@click.option(
+    '--log-level',
+    default='INFO',
+    type=click.Choice(list(__LOG_LEVELS__.keys()), case_sensitive=False),
+    help='Determines the verbosity of script output',
+)
+@click.option(
+    '--extensions',
+    default=["cpp", "cc", "c", "h", "hpp"],
+    type=str, multiple=True,
+    help='What file extensions to consider',
+)
+@click.option(
+    '--known-failure',
+    type=str, multiple=True,
+    help='What paths are known to fail',
+)
+@click.option(
+    '--skip-dir',
+    type=str,
+    multiple=True,
+    help='Skip a specific sub-directory from checks',
+)
+@click.argument('dirs',
+                type=click.Path(exists=True, file_okay=False, resolve_path=True), nargs=-1)
+def main(log_level, extensions, dirs, known_failure, skip_dir):
+    coloredlogs.install(level=__LOG_LEVELS__[log_level],
+                        fmt='%(asctime)s %(levelname)-7s %(message)s')
+
+    if not dirs:
+        logging.error("Please provide at least one directory to scan")
+        sys.exit(1)
+
+    if not extensions:
+        logging.error("Need at  least one extension")
+        sys.exit(1)
+
+    checker = OrphanChecker()
+    for k in known_failure:
+        checker.AddKnownFailure(k)
+
+    # ensure all GN data is loaded
+    for directory in dirs:
+        for name in Path(directory).rglob("BUILD.gn"):
+            checker.AppendGnData(name)
+
+    skip_dir = set(skip_dir)
+
+    # Go through all files and check for orphaned (if any)
+    extensions = set(extensions)
+    for directory in dirs:
+        for path, dirnames, filenames in os.walk(directory):
+            if any([s in path for s in skip_dir]):
+                continue
+            for f in filenames:
+                full_path = Path(os.path.join(path, f))
+                if not full_path.suffix or full_path.suffix[1:] not in extensions:
+                    continue
+                checker.Check(directory, full_path)
+
+    if checker.failures:
+        logging.warning("%d files not known to GN (%d fatal)", checker.failures, checker.fatal_failures)
+
+    if checker.known_failures != checker.found_failures:
+        not_failing = checker.known_failures - checker.found_failures
+        logging.warning("NOTE: %d failures are not found anymore:", len(not_failing))
```

Comment:
Would it be worthwhile to make this a fatal failure, so adding a previously good file to a gn build forces the dev to update the CI list?

Reply:
sure, I will update it. I added this late in the process and it seemed useful.